snakemake/utils.py

```
@@ -111,32 +111,42 @@
+            recordlist = []
+            for i, record in enumerate(records):
+                # Exclude NULL values
+                record = {k: v for k, v in record.items() if not pd.isnull(v)}
+                try:
+                    if set_default:
+                        DefaultValidator(schema, resolver=resolver).validate(record)
+                        recordlist.append(record)
+                    else:
+                        jsonschema.validate(record, schema, resolver=resolver)
+                except jsonschema.exceptions.ValidationError as e:
+                    raise WorkflowError(f"Error validating row {i} of data frame.", e)
```

Ruff (0.8.2)

Within an `except` clause, raise exceptions with `raise ... from err` or `raise ... from None` to distinguish them from errors in exception handling

(B904)